### ISSUE TITLE
Concurrent mode start failed with `uvicorn`

### ISSUE THREAD

**Issue by Ox0400** [PrefectHQ/prefect#13279]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

```python
Process SpawnProcess-7:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/local/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.10/site-packages/uvicorn/_subprocess.py", line 76, in subprocess_started
    target(sockets=sockets)
  File "/usr/local/lib/python3.10/site-packages/uvicorn/server.py", line 61, in run
    return asyncio.run(self.serve(sockets=sockets))
  File "/usr/local/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/usr/local/lib/python3.10/site-packages/uvicorn/server.py", line 68, in serve
    config.load()
  File "/usr/local/lib/python3.10/site-packages/uvicorn/config.py", line 479, in load
    self.loaded_app = self.loaded_app()
  File "/usr/local/lib/python3.10/site-packages/prefect/server/api/server.py", line 667, in create_app
    ui_app = create_ui_app(ephemeral)
  File "/usr/local/lib/python3.10/site-packages/prefect/server/api/server.py", line 408, in create_ui_app
    create_ui_static_subpath()
  File "/usr/local/lib/python3.10/site-packages/prefect/server/api/server.py", line 385, in create_ui_static_subpath
    copy_directory(prefect.__ui_static_path__, static_dir)
  File "/usr/local/lib/python3.10/site-packages/prefect/server/api/server.py", line 228, in copy_directory
    shutil.rmtree(destination)
  File "/usr/local/lib/python3.10/shutil.py", line 731, in rmtree
    onerror(os.rmdir, path, sys.exc_info())
  File "/usr/local/lib/python3.10/shutil.py", line 729, in rmtree
    os.rmdir(path)
OSError: [Errno 39] Directory not empty: '/usr/local/lib/python3.10/site-packages/prefect/server/ui_build/assets'
```

### Reproduction

```python3
export WEB_CONCURRENCY=10
prefect server start
``
```

### Versions

```Text
2.18.0
```

**Comment by mj0nez** [comment]
I had the same problem. After digging a bit I think the problem is that the static assets are copied in parallel when `uvicorn` uses the app factory to create its worker processes:
see [create_ui_app](https://github.com/PrefectHQ/prefect/blob/6e9adb80bae4eb5d6aedff2c63bb5c844fdf615a/src/prefect/server/api/server.py#L639) -> [create_ui_static_subpath](https://github.com/PrefectHQ/prefect/blob/6e9adb80bae4eb5d6aedff2c63bb5c844fdf615a/src/prefect/server/api/server.py#L358) -> [copy_directory](https://github.com/PrefectHQ/prefect/blob/6e9adb80bae4eb5d6aedff2c63bb5c844fdf615a/src/prefect/server/api/server.py#L235)

You can verify this by disabling the UI with `PREFECT_UI_ENABLED=false` or by providing the the assets before the server starts. In this case the server starts without errors because no process tries to copy the static assets.

**Comment by Ox0400** [comment]
> I had the same problem with version 3.0.2. After digging a bit I think the problem is that the static assets are copied in parallel when `uvicorn` uses the app factory to create its worker processes: see [create_ui_app](https://github.com/PrefectHQ/prefect/blob/6e9adb80bae4eb5d6aedff2c63bb5c844fdf615a/src/prefect/server/api/server.py#L639) -> [create_ui_static_subpath](https://github.com/PrefectHQ/prefect/blob/6e9adb80bae4eb5d6aedff2c63bb5c844fdf615a/src/prefect/server/api/server.py#L358) -> [copy_directory](https://github.com/PrefectHQ/prefect/blob/6e9adb80bae4eb5d6aedff2c63bb5c844fdf615a/src/prefect/server/api/server.py#L235)
> 
> You can verify this by disabling the UI with `PREFECT_UI_ENABLED=false` or by providing the assets before the server starts. In this case the server starts without errors because no process tries to copy the static assets.

Thanks to you, it's helpful!

**Comment by desertaxle** [comment]
I think this was fixed by #22363, which added per-bundle locking and a post-lock cache-key recheck while closing the newer duplicate #22362, so I'm going to close this issue. Please open a new issue if you are still experiencing this error.
